Example failing build: https://github.com/apache/airflow/actions/runs/5199718582/jobs/9378018845


```
=================================== FAILURES ===================================
_ test_get_safe_url[ javascript:alert(1)-http://localhost:8080/ javascript:alert(1)] _

mock_url_for = <MagicMock name='url_for' id='139816089007056'>
app = <Flask 'airflow.www.app'>, test_url = ' javascript:alert(1)'
expected_url = 'http://localhost:8080/ javascript:alert(1)'

    @pytest.mark.parametrize(
        "test_url, expected_url",
        [
            ("", "/home"),
            ("javascript:alert(1)", "/home"),
            (" javascript:alert(1)", "http://localhost:8080/ javascript:alert(1)"),
            ("http://google.com/", "/home"),
            ("google.com", "http://localhost:8080/google.com"),
            ("\\/google.com", "http://localhost:8080/\\/google.com"),
            ("//google.com", "/home"),
            ("\\/\\/google.com", "http://localhost:8080/\\/\\/google.com"),
            ("36539'%3balert(1)%2f%2f166", "/home"),
            (
                "http://localhost:8080/trigger?dag_id=test&origin=36539%27%3balert(1)%2f%2f166&abc=2",
                "/home",
            ),
            (
                "http://localhost:8080/trigger?dag_id=test_dag&origin=%2Ftree%3Fdag_id%test_dag';alert(33)//",
                "/home",
            ),
            (
                "http://localhost:8080/trigger?dag_id=test_dag&origin=%2Ftree%3Fdag_id%3Dtest_dag",
                "http://localhost:8080/trigger?dag_id=test_dag&origin=%2Ftree%3Fdag_id%3Dtest_dag",
            ),
        ],
    )
    @mock.patch("airflow.www.views.url_for")
    def test_get_safe_url(mock_url_for, app, test_url, expected_url):
        mock_url_for.return_value = "/home"
        with app.test_request_context(base_url="http://localhost:8080"):
>           assert get_safe_url(test_url) == expected_url
E           AssertionError: assert '/home' == 'http://local...ript:alert(1)/'
E             - http://localhost:8080/ javascript:alert(1)
E             + /home
```

--- a/airflow/www/views.py
+++ b/airflow/www/views.py
@@ -153,6 +153,27 @@ def sanitize_args(args: dict[str, str]) -> dict[str, str]:
     return {key: value for key, value in args.items() if not key.startswith("_")}
 
 
+# Following the release of https://github.com/python/cpython/issues/102153 in Python 3.8.17 and 3.9.17 on
+# June 6, 2023, we are adding extra sanitization of the urls passed to get_safe_url method to make it works
+# the same way regardless if the user uses latest Python patchlevel versions or not. This also follows
+# a recommended solution by the Python core team.
+#
+# From: https://github.com/python/cpython/commit/d28bafa2d3e424b6fdcfd7ae7cde8e71d7177369
+#
+#   We recommend that users of these APIs where the values may be used anywhere
+#   with security implications code defensively. Do some verification within your
+#   code before trusting a returned component part.  Does that ``scheme`` make
+#   sense?  Is that a sensible ``path``?  Is there anything strange about that
+#   ``hostname``?  etc.
+#
+# C0 control and space to be stripped per WHATWG spec.
+# == "".join([chr(i) for i in range(0, 0x20 + 1)])
+_WHATWG_C0_CONTROL_OR_SPACE = (
+    "\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c"
+    "\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f "
+)
+
+
 def get_safe_url(url):
     """Given a user-supplied URL, ensure it points to our web server."""
     if not url:
@@ -162,6 +183,8 @@ def get_safe_url(url):
     # potential XSS. (Similar to https://github.com/python/cpython/pull/24297/files (bpo-42967))
     if ";" in unquote(url):
         return url_for("Airflow.index")
+
+    url = url.lstrip(_WHATWG_C0_CONTROL_OR_SPACE)
 
     host_url = urlsplit(request.host_url)
     redirect_url = urlsplit(urljoin(request.host_url, url))

--- a/tests/www/views/test_views.py
+++ b/tests/www/views/test_views.py
@@ -193,7 +193,7 @@ def test_task_dag_id_equals_filter(admin_client, url, content):
     [
         ("", "/home"),
         ("javascript:alert(1)", "/home"),
-        (" javascript:alert(1)", "http://localhost:8080/ javascript:alert(1)"),
+        (" javascript:alert(1)", "/home"),
         ("http://google.com", "/home"),
         ("google.com", "http://localhost:8080/google.com"),
         ("\\/google.com", "http://localhost:8080/\\/google.com"),

